### PR TITLE
Added invite logic for local sign in 

### PR DESCRIPTION
**Pull request by farhatahmad**
## Testing Steps

Still a bit hard to test. You can manually generate an Invitation and try using it (will be testable next PR with emails)

### CODE REVIEW

File: app/javascript/components/home/HomePage.jsx [hunk 1]
```diff
@@ -27,8 +33,16 @@ export default function HomePage() {
...
   );
 
+  useEffect(() => {
+    document.cookie = `token=${inviteToken};path=/;`;
+  }, [inviteToken]);
+
   if (isLoading) return <Spinner />;
 
+  function showSignUp() {
+    return registrationMethod !== 'invite' || !!inviteToken;
```
Comment (scouillard) [inline app/javascript/components/home/HomePage.jsx:43]: Kind of hard to follow.
I would personally prefer a `===` operator instead and adding an extra clause.
SignUp will show up if the registration method is "Open" or "Approval" OR if the inviteToken (the URL param) is blank?

**Review by scouillard — APPROVED** [review]

File: app/javascript/components/home/HomePage.jsx (hunk 1 above)
Comment (farhatahmad) [reply app/javascript/components/home/HomePage.jsx:43]: Yeah the double bang is slightly annoying. It won't show Signup if the registration method is invite and there is no invite token. I'll add a comment here

File: app/controllers/api/v1/users_controller.rb [hunk 2]
```diff
@@ -97,12 +107,18 @@ def change_password
...
       end
 
       def change_password_params
         params.require(:user).permit(:old_password, :new_password)
       end
+
+      def valid_invite_token
+        return false if user_params[:invite_token].blank?
+
+        Invitation.exists?(email: user_params[:email], provider: current_provider, token: user_params[:invite_token])
```
Comment (KH-Amir-TN) [inline app/controllers/api/v1/users_controller.rb:120]: I believe this will have the validity check in the future?

File: app/controllers/api/v1/users_controller.rb [hunk 3]
```diff
@@ -26,12 +26,22 @@ def create
...
+
+        user = UserCreator.new(user_params: user_params.except(:invite_token), provider: current_provider, role: default_role).call
 
         # TODO: Add proper error logging for non-verified token hcaptcha
         return render_error errors: user.errors.to_a if hcaptcha_enabled? && !verify_hcaptcha(response: params[:token])
 
         if user.save
+          # Delete invitation (ignore whether it exists or not)
+          if registration_method == SiteSetting::REGISTRATION_METHODS[:invite]
+            Invitation.delete_by(email: user_params[:email], provider: current_provider,
```
Comment (KH-Amir-TN) [inline app/controllers/api/v1/users_controller.rb:41]: This isn't a problem here, but it's recommended to avoid the use of `delete_by` unless required since it bypasses dependencies,

File: app/javascript/components/home/HomePage.jsx [hunk 4]
```diff
@@ -27,8 +33,16 @@ export default function HomePage() {
     [currentUser.signed_in],
   );
 
+  useEffect(() => {
+    document.cookie = `token=${inviteToken};path=/;`;
```
Comment (KH-Amir-TN) [inline app/javascript/components/home/HomePage.jsx:37]: I suspect that this will always create the cookie,

File: app/javascript/components/home/HomePage.jsx (hunk 1 above)
Comment (KH-Amir-TN) [reply app/javascript/components/home/HomePage.jsx:43]: Agreed,
Besides, in JS there's no need to cast a value to a boolean for conditional checks,
`!!inviteToken` should be the same as `inviteToken` in a condition,

File: app/javascript/components/home/HomePage.jsx (hunk 4 above)
Comment (KH-Amir-TN) [reply app/javascript/components/home/HomePage.jsx:37]: Why we don't host this logic on signup where it will check the validity of the request,
It will further set any required cookies and query params for the signup request,

Comment (KH-Amir-TN) [reply app/javascript/components/home/HomePage.jsx:37]: So the invitation URL is a `.../signup?inviteToken=xxx`

File: app/models/invitation.rb [hunk 5]
```diff
@@ -5,5 +5,5 @@ class Invitation < ApplicationRecord
 
   validates :email, presence: true, uniqueness: { scope: :provider }
   validates :provider, presence: true
-  validates :token, presence: true, uniqueness: true
+  validates :token, uniqueness: true
```
Comment (KH-Amir-TN) [inline app/models/invitation.rb:8]: Is this removed beacuse it's guranteed to be present?
